test/src/View/MoveToolControllerTest.cpp

```diff
@@ -0,0 +1,146 @@
+            MOCK_METHOD0(mockDoCancelMove, void());
+        public:
+            using MoveToolController::MoveInfo;
+            using RestrictedDragPolicy::DragResult;
+            using RestrictedDragPolicy::DR_Continue;
+        };
+        
+        TEST(MoveToolControllerTest, testMoveWithSnapUp) {
+            using namespace ::testing;
+            using ::testing::InSequence;
+
+            const Renderer::Camera::Viewport viewport(-200, -200, 400, 400);
+            Renderer::PerspectiveCamera camera(90.0f, 0.1f, 500.0f, viewport, Vec3f(0.0f, 0.0f, 100.0f), Vec3f::NegZ, Vec3f::PosY);
+            
+            const Grid grid(4); // Grid size 16
+            MockMoveToolController controller(grid);
+            
+            InputState inputState(0, 0);
+            inputState.mouseDown(MouseButtons::MBLeft);
+            
+            const Vec3 origin = Vec3(camera.position());
+            inputState.setPickRequest(PickRequest(Ray3(origin, Vec3::NegZ), camera));
+
+            EXPECT_CALL(controller, mockDoStartMove(Ref(inputState))).Times(1).WillOnce(Return(MockMoveToolController::MoveInfo(Vec3::Null)));
+            controller.startMouseDrag(inputState);
+            
+            inputState.mouseMove(9, 0, 9, 0);
+            inputState.setPickRequest(PickRequest(Ray3(origin, (Vec3(9.0, 0.0, 0.0) - origin).normalized()), camera));
+            
+            EXPECT_CALL(controller, mockDoMove(Ref(inputState), Vec3(0.0, 0.0, 0.0), Vec3(16.0, 0.0, 0.0))).Times(1).WillOnce(Return(MockMoveToolController::DR_Continue));
+            controller.mouseDrag(inputState);
+            
+            inputState.mouseUp(MouseButtons::MBLeft);
+            EXPECT_CALL(controller, mockDoEndMove(Ref(inputState)));
+            controller.endMouseDrag(inputState);
+        }
+        
+        TEST(MoveToolControllerTest, testMoveAfterZeroVerticalMove) {
+            // see https://github.com/kduske/TrenchBroom/issues/1529
+            
+            using namespace ::testing;
+            using ::testing::InSequence;
+            
+            const Renderer::Camera::Viewport viewport(-200, -200, 400, 400);
+            Renderer::PerspectiveCamera camera(90.0f, 0.1f, 500.0f, viewport, Vec3f(0.0f, 0.0f, 100.0f), Vec3f::NegZ, Vec3f::PosY);
+            
+            const Grid grid(4); // Grid size 16
+            MockMoveToolController controller(grid);
+            
+            InputState inputState(0, 0);
+            inputState.mouseDown(MouseButtons::MBLeft);
+            
+            const Vec3 origin = Vec3(camera.position());
+            inputState.setPickRequest(PickRequest(Ray3(origin, Vec3::NegZ), camera));
+            
+            EXPECT_CALL(controller, mockDoStartMove(Ref(inputState))).Times(1).WillOnce(Return(MockMoveToolController::MoveInfo(Vec3::Null)));
+            controller.startMouseDrag(inputState);
+            
+            // nothing will happen due to grid snapping
```

Comment:
Add:
`EXPECT_CALL(controller, mockDoMove(_,_,_)).Times(0);`
to make the test fail if the controller actually receives a move.
Currently it will just print a warning.